File: ely/contexts.py
"""
Context management — persistent, customizable contexts stored as markdown files.

Contexts are stored in:
  - ~/.ely/contexts/<name>.md   (user-level, global)
  - ./.ely/contexts/<name>.md   (project-level, overrides user)

Each context file uses YAML frontmatter:
  ---
  name: code
  description: Mode développement
  ---
  Prompt text here...

Active context is persisted in ~/.ely/context.json
"""

import os


def _contexts_dirs() -> list[str]:
    """Return context search directories, project first (higher priority)."""
    from .config import get_ely_dir
    dirs = []
    project = os.path.join(os.getcwd(), ".ely", "contexts")
    if os.path.isdir(project):
        dirs.append(project)
    user = os.path.join(get_ely_dir(), "contexts")
    if os.path.isdir(user):
        dirs.append(user)
    return dirs


def _ensure_defaults():
    """Create default contexts if they don't exist."""
    from .config import get_ely_dir
    user_dir = os.path.join(get_ely_dir(), "contexts")
    os.makedirs(user_dir, exist_ok=True)

    defaults = {
        "default": {
            "description": "Mode général — terminal et tâches courantes",
            "prompt": "Tu es dans un terminal. L'utilisateur travaille dans le répertoire courant.",
        },
        "code": {
            "description": "Mode développement — exploration et écriture de code",
            "prompt": "L'utilisateur travaille sur du code. Utilise read_file, grep, et list_directory pour explorer la codebase avant de proposer des changements. Lis les fichiers avant de les modifier.",
        },
        "sysadmin": {
            "description": "Mode administration système — commandes shell",
            "prompt": "L'utilisateur fait de l'administration système. Utilise bash pour les commandes, mais sois prudent avec les commandes destructives. Vérifie toujours avant d'exécuter.",
        },
        "research": {
            "description": "Mode recherche — web et documentation",
            "prompt": "L'utilisateur fait de la recherche. Utilise les fichiers du workspace, web_search et web_fetch pour trouver des informations récentes. Synthétise et cite tes sources.",
        },
    }

    for name, info in defaults.items():
        path = os.path.join(user_dir, f"{name}.md")
        if not os.path.isfile(path):
            content = f"---\nname: {name}\ndescription: {info['description']}\n---\n\n{info['prompt']}"
            with open(path, "w") as f:
                f.write(content)


def _parse_context_file(path: str) -> dict | None:
    """Parse a context .md file. Returns {name, description, prompt} or None."""
    try:
        with open(path) as f:
            content = f.read()
    except Exception:
        return None

    meta = {}
    body = content

    if content.startswith("---"):
        parts = content.split("---", 2)
        if len(parts) >= 3:
            try:
                import yaml
                meta = yaml.safe_load(parts[1]) or {}
            except Exception:
                for line in parts[1].strip().split("\n"):
                    if ":" in line:
                        k, v = line.split(":", 1)
                        meta[k.strip()] = v.strip()
            body = parts[2].strip()

    return {
        "name": meta.get("name", os.path.basename(path)[:-3]),
        "description": meta.get("description", ""),
        "prompt": body,
    }


def list_contexts() -> list[dict]:
    """List all available contexts with metadata."""
    _ensure_defaults()
    seen = set()
    result = []

    for d in _contexts_dirs():
        if not os.path.isdir(d):
            continue
        for f in sorted(os.listdir(d)):
            if f.endswith(".md") and f[:-3] not in seen:
                info = _parse_context_file(os.path.join(d, f))
                if info:
                    result.append(info)
                    seen.add(f[:-3])

    return sorted(result, key=lambda c: c["name"])


def get_context(name: str) -> dict | None:
    """Get a context by name. Returns {name, description, prompt} or None."""
    _ensure_defaults()
    for d in _contexts_dirs():
        path = os.path.join(d, f"{name}.md")
        if os.path.isfile(path):
            return _parse_context_file(path)
    return None


def get_context_prompt(name: str) -> str:
    """Get just the prompt text for a context."""
    ctx = get_context(name)
    return ctx["prompt"] if ctx else ""


def create_context(name: str, description: str, prompt: str) -> str:
    """Create a custom context in the user directory. Returns path."""
    from .config import get_ely_dir
    user_dir = os.path.join(get_ely_dir(), "contexts")
    os.makedirs(user_dir, exist_ok=True)

    content = f"---\nname: {name}\ndescription: {description}\n---\n\n{prompt}"
    path = os.path.join(user_dir, f"{name}.md")
    with open(path, "w") as f:
        f.write(content)
    return path


def _active_context_file() -> str:
    from .config import get_ely_dir
    return os.path.join(get_ely_dir(), "context.json")


def save_active_context(name: str):
    """Persist the active context name."""
    import json
    os.makedirs(os.path.dirname(_active_context_file()), exist_ok=True)
    with open(_active_context_file(), "w") as f:
        json.dump({"context": name}, f)


def load_active_context() -> str:
    """Load the persisted active context. Returns 'default' if not set."""
    import json
    path = _active_context_file()
    if os.path.isfile(path):
        try:
            with open(path) as f:
                data = json.load(f)
            return data.get("context", "default")
        except Exception:
            pass
    return "default"


def delete_context(name: str) -> bool:
    """Delete a custom context. Cannot delete built-in defaults."""
    if name in ("default", "code", "sysadmin", "research"):
        return False
    for d in _contexts_dirs():
        path = os.path.join(d, f"{name}.md")
        if os.path.isfile(path):
            os.remove(path)
            return True
    return False

File: ely_rust/src/contexts.rs
use std::path::PathBuf;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Context {
    pub name: String,
    pub description: String,
    pub prompt: String,
}

fn context_dirs(ely_dir: &PathBuf) -> Vec<PathBuf> {
    let mut dirs = vec![std::env::current_dir().unwrap_or_default().join(".ely").join("contexts")];
    dirs.push(ely_dir.join("contexts"));
    dirs.into_iter().filter(|d| d.is_dir()).collect()
}

pub fn ensure_defaults(ely_dir: &PathBuf) {
    let user_dir = ely_dir.join("contexts");
    std::fs::create_dir_all(&user_dir).ok();
    let defaults = [
        ("default", "Mode général — terminal et tâches courantes", "Tu es dans un terminal. L'utilisateur travaille dans le répertoire courant."),
        ("code", "Mode développement — exploration et écriture de code", "L'utilisateur travaille sur du code. Utilise read_file, grep, et list_directory pour explorer la codebase."),
        ("sysadmin", "Mode administration système — commandes shell", "L'utilisateur fait de l'administration système. Utilise bash pour les commandes, sois prudent."),
        ("research", "Mode recherche — web et documentation", "L'utilisateur fait de la recherche. Utilise web_search et web_fetch."),
    ];
    for (name, desc, prompt) in &defaults {
        let path = user_dir.join(format!("{}.md", name));
        if !path.exists() {
            let content = format!("---\nname: {}\ndescription: {}\n---\n\n{}", name, desc, prompt);
            std::fs::write(&path, content).ok();
        }
    }
}

pub fn list_contexts(ely_dir: &PathBuf) -> Vec<Context> {
    ensure_defaults(ely_dir);
    let mut seen = std::collections::HashSet::new();
    let mut result = Vec::new();
    for d in context_dirs(ely_dir) {
        if let Ok(entries) = std::fs::read_dir(&d) {
            for e in entries.flatten() {
                let name = e.file_name().to_string_lossy().into_owned();
                if name.ends_with(".md") {
                    let context_name = name.trim_end_matches(".md").to_string();
                    if seen.insert(context_name.clone()) {
                        if let Some(ctx) = parse_context(&e.path()) {
                            result.push(ctx);
                        }
                    }
                }
            }
        }
    }
    result.sort_by(|a, b| a.name.cmp(&b.name));
    result
}

pub fn get_context(ely_dir: &PathBuf, name: &str) -> Option<Context> {
    ensure_defaults(ely_dir);
    for d in context_dirs(ely_dir) {
        let path = d.join(format!("{}.md", name));
        if path.exists() {
            return parse_context(&path);
        }
    }
    None
}

pub fn get_context_prompt(ely_dir: &PathBuf, name: &str) -> String {
    get_context(ely_dir, name).map(|c| c.prompt).unwrap_or_default()
}

pub fn create_context(ely_dir: &PathBuf, name: &str, description: &str, prompt: &str) -> PathBuf {
    let dir = ely_dir.join("contexts");
    std::fs::create_dir_all(&dir).ok();
    let content = format!("---\nname: {}\ndescription: {}\n---\n\n{}", name, description, prompt);
    let path = dir.join(format!("{}.md", name));
    std::fs::write(&path, content).ok();
    path
}

pub fn delete_context(ely_dir: &PathBuf, name: &str) -> bool {
    if ["default", "code", "sysadmin", "research"].contains(&name) { return false; }
    for d in context_dirs(ely_dir) {
        let path = d.join(format!("{}.md", name));
        if path.exists() { return std::fs::remove_file(&path).is_ok(); }
    }
    false
}

pub fn save_active_context(ely_dir: &PathBuf, name: &str) {
    let path = ely_dir.join("context.json");
    if let Ok(json) = serde_json::to_string(&serde_json::json!({"context": name})) {
        std::fs::write(&path, json).ok();
    }
}

pub fn load_active_context(ely_dir: &PathBuf) -> String {
    let path = ely_dir.join("context.json");
    std::fs::read_to_string(&path).ok()
        .and_then(|s| serde_json::from_str::<serde_json::Value>(&s).ok())
        .and_then(|v| v["context"].as_str().map(|s| s.to_string()))
        .unwrap_or_else(|| "default".into())
}

fn parse_context(path: &PathBuf) -> Option<Context> {
    let content = std::fs::read_to_string(path).ok()?;
    let (meta, body) = if content.starts_with("---") {
        let parts: Vec<&str> = content.splitn(3, "---").collect();
        if parts.len() >= 3 {
            let meta: HashMap<String, String> = serde_yaml::from_str(parts[1]).unwrap_or_default();
            (meta, parts[2].trim().to_string())
        } else {
            (HashMap::new(), content)
        }
    } else {
        (HashMap::new(), content)
    };
    Some(Context {
        name: meta.get("name").cloned().unwrap_or_else(|| path.file_stem().unwrap_or_default().to_string_lossy().into()),
        description: meta.get("description").cloned().unwrap_or_default(),
        prompt: body,
    })
}

use std::collections::HashMap;
